src/net_processing.cpp

```diff
@@ -906,7 +907,11 @@ void PeerLogicValidation::ReattemptInitialBroadcast(CScheduler& scheduler) const
 }
 
 void PeerLogicValidation::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTime) {
+    int misbehavior{0};
     {
+        PeerRef peer = GetPeerRef(nodeid);
+        assert(peer != nullptr);
+        misbehavior = WITH_LOCK(peer->m_misbehavior_mutex, return peer->nMisbehavior);
```

Comment:
For functions that access multiple fields under the same lock, we'll need to hold the lock externally to the struct anyway (see `Misbehaving()` for example, which accesses both `m_misbehavior_score` and `m_should_discourage` while holding `m_misbehavior_score`). So I could add setters/getters which handle the locking, but the locks would still need to be public and it wouldn't help with encapsulating `Peer`.

I like that `Peer` is a struct with all public members and no internal logic, and I also think the `WITH_LOCK` macro is pretty nifty so don't mind this pattern. If others also have a strong distaste for the WITH_LOCK pattern, I can look at adding getter methods.

Reply:
I'm fond of WITH_LOCK.  Especially if we're going the way of many individual locks, it's much quicker this way to see what's locking what.

This does give me flashbacks to Java though... `Button button = new Button("button");`

Reply:
@sdaftuar This was my high-level feedback as well. I strongly prefer encapsulation as it's really really hard to screw up encapsulated locking, but it's a burden when multiple operations could otherwise be handled with a single lock. But as this work/PR is intended to break up locking that has grown unwieldy, I can see why @jnewbery went for this approach.

Also, just in terms of evolution of thinking... I have always preferred encapsulated locks because of the security guarantees. But I'm less against exposed mutexes these days due to our annotations and sanitizers.

Reply:
It seems like there are two different points here:

- "This is a mouthful" (@sdaftuar) seems like a stylistic comment. I actually like the fact that the WITH_LOCK macro is explicit about the lock it's taking.
- "I strongly prefer encapsulation as it's really really hard to screw up encapsulated locking" (@theuni) is a comment about locking design. Cory, correct me if I'm wrong, but I think your distaste here is informed by how cs_main has been (mis)used in the past, where it's locked the mutex for huge scopes unnecessarily. With lock annotations and fine-grained mutexes, I don't see a problem with exposing those mutexes outside the class/struct.

Reply:
My primary concern is that 5 years from now someone changes some code that interacts with a Peer and forgets to grab a lock. With encapsulated locking, that _can't_ happen.

But I'm now satisfied that even without the encapsulation, we have safeguards to prevent that.